### ISSUE TITLE
With Horus moving tags, non-Tanaka Corpus sentences are getting mislabelled.

### ISSUE THREAD

**Issue by ckjpn** [Tatoeba/tatoeba2#2595]
https://tatoeba.org/eng/sentences/show/1439987

<img width="1120" alt="Screen Shot 2020-12-16 at 16 57 16" src="https://user-images.githubusercontent.com/7934273/102320638-ec423f00-3fbf-11eb-9c29-0482ad37ed65.png">

Maybe Tanaka Corpus tags should be removed and another way to indicate sentences are from the Tanaka Corpus should be considered.

**Comment by trang** [comment]
It's not a bug in my opinion.

If a sentence from the Tanaka Corpus gets deleted because it was orphan and someone in Tatoeba created a duplicated, why should the tag be removed? The tag is an information that there was a corpus called Tanaka Corpus and in the sentence was part of that corpus.

In your example, the original sentence was `We have nothing to complain of.`. This sentence was part of the Tanaka Corpus. It was changed to `We have nothing to complain about.` and it became a duplicate of one of your sentences.

I guess one could argue that if a sentence from the Tanaka Corpus gets corrected in Tatoeba, the tag "Tanaka Corpus" is no longer relevant but this can only be decided case by case. In this case, I think the tag is still relevant because there has only been a minor correction.

I'm closing this but feel free to re-open if there's another perspective to thsi issue that I'm not seeing.

**Comment by ckjpn** [comment]
I disagree with your opinion on this.

Since the "about" version was a different sentence, it was not a Tanaka Corpus sentences, so tagging it as such is not true and not valid.  I think you should at least make an attempt at truthfully tagging sentences, otherwise why tag them with this tag at all.

Note that "complain of" and "complain about" have different meanings.

Also, if the correct sentence was not part of the imported Tanaka Corpus, why should it be called part of the Tanaka Corpus, after someone changes it and then Horus moves the tag.  This is illogical.  

I hope that you would reconsider your position on this and that "Tanaka Corpus" tags accurately are only on sentences that were imported with the Tanaka Corpus.

**Comment by AndiPersti** [comment]
I tend to agree with @ckjpn.

But I also think this issue isn't only about the tag for the Tanaka Corpus. Here's a list of all the tags that Horus moved since November 15 (the oldest Horus log entry we still have):

id    | name          
-|-      
5 | ambiguous           
7 | OK                  
155 | sign                
189 | colloquial       
233 | by Voltaire         
407 | proverb             
527 | time                
561 | @&#xfeff;change             
714 | female speaker   
841 | @&#xfeff;check              
945 | @&#xfeff;change punctuation 
1238 | 8 to 10 words       
2328 | comment on sentence 
4015 | by Astrid Lindgren  
4970 | Tanaka Corpus       
5296 | female name         
7770 | List 907            
10794 | @&#xfeff;please adopt       
11602 | Kipo                
11671 | Sénèque             
11683 | Pije la

I'd say that at least all the `@...` tags shouldn't be auto-merged either because the reason why a sentence was tagged with one of them probably isn't valid any more after the merging operation (e.g. some adds a punctuation so `@change punctuation` becomes outdated).
See also https://github.com/Tatoeba/tatoeba2/issues/1830#issuecomment-583591293

**Comment by trang** [comment]
> Since the "about" version was a different sentence, it was not a Tanaka Corpus sentences, so tagging it as such is not true and not valid.

The "about" version was created indeed as a separate sentence, but it was also the result of editing a Tanaka Corpus sentence to make it better.

In the [discussion on the sentence #318670](https://tatoeba.org/eng/sentences/show/318670), the community came to the conclusion that `complain of` should be changed to `complain about`. Now it happens that this correction led the sentence to become a duplicate, and it got deleted because it was orphan. But in my opinion it can still be considered part of the Tanaka Corpus. Of course it's not part of the exact corpus that was released by professor Tanaka many years ago, but it's still part of the project that he initiated.

If we remove the "Tanaka Corpus" tag, then what happens if someone wants, for some reason, to extract only sentences from the Tanaka Corpus? They would get the Japanese sentence ([https://tatoeba.org/eng/sentences/show/85043](#85043)), but then they would get no English translation for it because the English sentence is no longer tagged.

If the Japanese sentence was deleted or if it was unlinked, then sure, the tag could be removed from the English sentence. But otherwise, it is stilled tied to the Tanaka Corpus because it is a translation of a Japanese sentence from the Tanaka Corpus.

> I'd say that at least all the `@...` tags shouldn't be auto-merged either because the reason why a sentence was tagged with one of them probably isn't valid any more after the merging operation (e.g. some adds a punctuation so `@change punctuation` becomes outdated).

To me the problem doesn't come from Horus moving the tags. I think Horus should always move the tags and shouldn't have to ask itself whether a tag is still valid or not when merging.

The problem is what we do with tags when a sentence is modified. It's a problem that is independant from deduplication. Sometimes, the modification of a sentence can invalidate some of the tags, regardless if the sentence becomes a duplicate or not after the modification. 

If we have a way to detect automatically all tags that become invalid when the sentence is modified, then we should automatically remove them upon modification of the sentence. In which case, there would be no problem with moving wrong tags because there would be no wrong tag. Otherwise, we can only rely on advanced contributors to do some clean up work on the tags of sentences that have been modified.

**Comment by ckjpn** [comment]
> Now it happens that this correction led the sentence to become a duplicate, and it got deleted because it was orphan. But in my opinion it can still be considered part of the Tanaka Corpus.

I think you are wrong about this.

The original Tanaka Corpus sentences had "complain of" which, though perhaps a old-fashioned, was a likely valid sentence, even though AlanF_US went ahead and changed it.

However, even in the case where a sentence was totally wrong and corrected, such a sentence was not part of the original public domain Tanaka Corpus that was imported into the Tatoeba Corpus. It should be part of the CC-BY Tatoeba Corpus.

If you are going to tag sentences as "Tanaka Corpus," then they should actually be Tanaka Corpus sentences.  Moving the tags to another sentence created by a Tatoeba Project member is misleading, and inaccurate.

**Comment by trang** [comment]
> However, even in the case where a sentence was totally wrong and corrected, such a sentence was not part of the original public domain Tanaka Corpus that was imported into the Tatoeba Corpus. It should be part of the CC-BY Tatoeba Corpus.

If a sentence is in Tatoeba, it's in the Tatoeba Corpus no matter what tag it has... What makes you think that the "Tanaka Corpus" tag excludes the sentence from being in the Tatoeba Corpus?

> If you are going to tag sentences as "Tanaka Corpus," then they should actually be Tanaka Corpus sentences. Moving the tags to another sentence created by a Tatoeba Project member is misleading, and inaccurate.

If we disagree on what the tag means, we will of course disagree on whether or not a sentence should have the tag.

You seem to insist that this tag means "This is part of the original public domain Tanaka Corpus". But this tag, as we use it today, does *not* mean that.

It was added upon Alan's request, to be able to distinguish sentences that originate from the Tanaka Corpus and should be handled with extra care (see #2098). Based on that, I think the tag is okay because the sentence was not changed to something completely different. It still has roots from the Tanaka Corpus. It was born in the Tanaka Corpus, then someone saw it in Tatoeba and thought "it doesn't sound completely right" and it was modified to sound better. And it is still connected to the Japanese version, which has not been modified.

If the tag is supposed to mean "sentence part of the original public domain Tanaka Corpus", then many sentences are wrongly labeled because many of these sentences got corrected over the years within Tatoeba and no longer match what was in the original corpus. In fact, the original public domain Tanaka Corpus itself has been corrected and cleaned up even before it was imported into Tatoeba so, at the time of the import, we no longer had all the original sentences.

In any case, the problem here, I think has drifted far away from Horus. This is more an issue with the lack of description for what a tag means.

**Comment by ckjpn** [comment]
Perhaps you could have Horus rename the "Tanaka Corpus" tag as it's moving it to "Similar to what was imported as part of the Tanaka Corpus" or something much shorter and more concise that means the same thing.

The way it is now is inaccurate and doesn't really help anybody now or researchers in the future.

I think the easier solution would be to not have Horus move the "Tanaka Corpus" tag to a sentence that was not imported as part of the Tanaka Corpus.

**Comment by trang** [comment]
Again: we're not going to involve Horus in this. The issue of tags becoming invalid can happen to any sentence that is modified, not just sentences that become duplicate. It should therefore be resolved upon modification of the sentence. There may be tags for which this can be done automatically, but otherwise it needs to be done manually. For the "Tanaka Corpus" tag, I don't think it can be resolved automatically.

The fact that a tag is accurate or not depends on the definition of the tag. As of today, we don't have a way to add a description to a tag so people can interpret a same tag in different ways, making it inaccurate to them. Generally speaking, every tag is subject to ambiguity and having the possibility to add a description to tags would be extremely helpful in that regards.

You interpreted the tag "Tanaka Corpus" as "sentence from the original public domain corpus released by professor Tanaka", I get that. But again, it was not the intended meaning. The intended meaning is that the tag refers to sentences that originated from the Tanaka corpus. Please understand that this corpus was heavily edited after its original release and the edited versions were still referred as "Tanaka Corpus". What we've been calling "Tanaka Corpus" is not a fixed corpus released at a specific date in time. It's an evolving corpus that used to be hosted at WWWJDIC but is now hosted in Tatoeba and continues to be maintained in Tatoeba. The tag "Tanaka Corpus" refers to the sentences part of the maintained version. If someone wants to get the original version of the Tanaka Corpus, they can download it from http://www.edrdg.org/wiki/index.php/Tanaka_Corpus.

**Comment by ckjpn** [comment]
It looks like you're not going to be convinced.  However, I think moving the Tanaka Corpus tag to a non-Tanaka Corpus sentence is doing a disservice to whoever uses the Tatoeba Corpus.

**Comment by trang** [comment]
If instead of using a tag, we put these sentences in a list named "Tanaka Corpus", would that make any difference to you?

**Comment by ckjpn** [comment]
That would seem more logical, wouldn't it?  I assume Horus doesn't move list designations along with what it does.

You could set the list to the new "public" which would so to everyone on each page, just like tags do.

**Comment by trang** [comment]
Horus moves list designations, just like tags, favorites and other things.
Cf. https://github.com/Tatoeba/horus/blob/master/dedup/management/commands/deduplicate.py#L395

Having a public list is virtually the same as having a tag, with the only difference of how and where the information is displayed.

Knowing that, do you still feel that putting the sentence in a list is more appropriate? If yes, then I'd be interested to understand why exactly. This would shed some light on the conceptual differences between lists and tags.

**Comment by ckjpn** [comment]
If sentences added to the Tatoeba Corpus after the import of the Tanaka Corpus are added to a "Tanaka Corpus" list, then the same problem still exists, so it doesn't really help.

One advantage of a list would be that there could be no possibility of Corpus Maintainers and other Admins accidentally deleting a Tag (or list item), since if set properly, only the list owner, you, would be able to add or delete items.

It seems that you should be able to put in an "if" statement, saying "if the tag = 'Tanaka Corpus', then don't move it."  I haven't looked at the code at all, but I would assume that this would not be very difficult for a programmer familiar with the tatoeba code and CakePHP.

**Comment by trang** [comment]
> If sentences added to the Tatoeba Corpus after the import of the Tanaka Corpus are added to a "Tanaka Corpus" list, then the same problem still exists, so it doesn't really help.

Then we will keep the tag. 

I can definitely see advantages in using a list, but the new "public" visibility was not available back when #2098 was requested and I didn't think about it when Andi worked on effectively adding the tags, so we'll just keep what we have.

We may one day set up something to convert a tag into a list. The "Tanaka Corpus" tag is not the only tag that would be better as a list.

> It seems that you should be able to put in an "if" statement, saying "if the tag = 'Tanaka Corpus', then don't move it."

No, this is not a viable solution. Again: the question of keeping or not a tag has to be solved at the time when the sentence is modified, not at the time when it is merged.

If a sentence if a tag becomes invalid after a sentence is modified, then it should be removed after the modification. If a tag is still valid, then the tags should stay on that sentence, even if the sentence is merged into another ID.

**Comment by ckjpn** [comment]
> If a sentence if a tag becomes invalid after a sentence is modified, then it should be removed after the modification. If a tag is still valid, then the tags should stay on that sentence, even if the sentence is merged into another ID.

Does this mean that from time to time I should check the exported data and delete all Tanaka Corpus tags on sentences above the number of the last imported Tanaka Corpus sentence?

It still seems that any moderately-competent programmer should be able to prevent this problem from happening with an if statement. I'm not familiar enough with CakePHP to know.

**Comment by alanfgh** [comment]
@ckjpn, you have the following options:

1. Accept the situation the way it is.
2. Periodically delete tags as you suggested. Trang certainly won't stop you.
3. Learn CakePHP, write code, and submit a pull request.
4. Attempt to badger other programmers into doing it by continually posting about it, saying that it couldn't be hard to implement, and ignoring their objections (as well as the possibility that they're busy taking care of higher-priority issues).

For most people, solution 1 would make the most sense. I doubt many people care about whether a sentence has a "Tanaka Corpus" tag, given that the tag says little about the quality of the sentence. (It could have been correct when it was added or corrected later.)

For you, given that you have a lot of spare time to spend on Tatoeba, you could do either 2 or 3. Granted, 3 would only work if you could become a moderately competent CakePHP programmer. That may be beyond your skill set. However, learning CakePHP would free you to make the changes you want in general, so the investment in time could pay for itself.

However, approach 4, which seems to be your favorite, is wasting other people's time, annoying them, and keeping them from doing more important things.

**Comment by ckjpn** [comment]
Question to TRANG.

Does alfanfgh's opinion reflect yours?
Would you prefer that I not try to help the Tatoeba Project when I see problems like this?

**Comment by AndiPersti** [comment]
> It still seems that any moderately-competent programmer should be able to prevent this problem from happening with an if statement. 

As I understand it this issue is not about how we should change the code but whether we should change it at all.

**Comment by AndiPersti** [comment]
> I'm not familiar enough with CakePHP to know.

Just for your information: Horus is implemented in Python and has its own repository: https://github.com/Tatoeba/horus
So you wouldn't need to know CakePHP to change Horus.

**Comment by trang** [comment]
@ckjpn

Alan's opinion reflects partially my opinion.

1. You can indeed simply accept the situation the way it is and that may be the best thing you can do here. That doesn't mean that I would prefer you don't try to help, but if your way of helping is to just systematically remove tags on sentences that have an ID higher than the highest ID for the Tanaka Corpus, then that's not helping because you will for sure remove some tags "Tanaka Corpus" tags that are valid.
2. I will indeed not stop you from removing "Tanaka Corpus" tags periodically but as we have different opinions on which sentence should be tagged "Tanaka Corpus", it might be best you don't remove these tags at all. You can, however, post a comment on the sentence to express that you believe the tag should be removed, then I can check and remove it if believe it should be removed too.
3. You are always welcome to learn to write code and submit pull requests, but in this specific case, if you would submit a pull request to add a "if" statement saying "if the tag = 'Tanaka Corpus', then don't move it", I wouldn't approve it. Again: I don't think the decision can be automated. A human has to evaluate if the sentence has been changed sufficiently to consider that it's a completely different sentence, and not a sentence that has only received minor improvement but kept the same meaning.
4. You are free to try and convince other developers to implement a solution for your problem with the Tanaka Corpus tag. But again: I will not agree with any "if" statement to automatically remove the tag by checking the IDs. I would however be happy to have descriptions for tags, so that we can have more details on what a tag means and when should a sentence be tagged with it. I would also be happy to have a feature for converting a tag into a list because the "Tanaka Corpus" would be better as a list than as a tag.

**Comment by alanfgh** [comment]
> You are free to try and convince other developers to implement a solution for your problem with the Tanaka Corpus tag

but if your attempt to "convince" involves: 

- repeatedly making the same request,
- ignoring the objections that are made in reponse,
- insisting on a particular approach in the code,
- demonstrating ignorance about the code,
- failing to actually learn the code or the languages in which it is written,
- disparaging the effort required to address the problem, and
- refusing to accept that developers have other priorities

then don't be surprised if you don't have much success.

**Comment by jiru** [comment]
I’d like to share my opinion on that.

Is there something we can currently call the "Tanaka Corpus", now that it has been merged into the Tatoeba Corpus with its sentences corrected or slightly modified? I don’t think so. The "Tanaka Corpus" tag is merely telling the origin of the sentences and crediting past work. From the day it was imported into Tatoeba, it became a derivative work. I can envision a tag category (#333) like "origin" and the tag would show "origin: Tanaka Corpus", or even a new field to track the origin of sentences as we perform new mass imports in the future (like #2256). From that perspective, the meaning of the current tag is a bit confusing, but so are many of our tags until more work is done.

Now when two sentences get merged into one, could the merged sentence have two origins? It doesn’t make much sense to me. It’s like having two owners. I think the origin of the sentence that gets removed by the merging process should disappear along with it.

To sum up, I think that merging of the tag "Tanaka Corpus" should be special-cased, but not until we have a way to programmatically know whether a particular tag is about origin or not.

> If we remove the "Tanaka Corpus" tag, then what happens if someone wants, for some reason, to extract only sentences from the Tanaka Corpus? They would get the Japanese sentence (https://tatoeba.org/eng/sentences/show/85043), but then they would get no English translation for it because the English sentence is no longer tagged.

It’s already impossible to extract sentences from the Tanaka Corpus that have been deleted on purpose (not because of merge). I don’t know if such deletions actually happened, but my point is that copying the tag onto the merged sentence is not a viable solution to the use case you’re describing. If we really want to keep track of the origin of deleted sentences (be it because of a merge or on purpose), I think this information should stay attached the deleted sentence, similarly to how comments are kept on deleted sentences.
